Holovibes/sources/data_struct/batch_input_queue.cc

```diff
@@ -121,70 +121,89 @@ void BatchInputQueue::stop_producer()
-    else
-        cudaXMemcpyAsync(new_frame_adress, input_frame, sizeof(char) * fd_.get_frame_size(), memcpy_kind);
+        // Static_cast to avoid overflow
+        char* const new_frame_adress =
+            data_.get() +
+            ((static_cast<size_t>(end_index_) * batch_size_ + curr_batch_counter_) * fd_.get_frame_size());
 
-    // No sync needed here, the host doesn't need to wait for the copy to
-    // end. Only the consumer needs to be sure the data is here before
-    // manipulating it.
+        if (device_ == Device::GPU)
+            cudaXMemcpyAsync(new_frame_adress,
+                             frames,
+                             sizeof(char) * fd_.get_frame_size() * frames_to_enqueue,
+                             memcpy_kind,
+                             batch_streams_[end_index_]);
+        else
+            cudaXMemcpyAsync(new_frame_adress,
+                             frames,
+                             sizeof(char) * fd_.get_frame_size() * frames_to_enqueue,
+                             memcpy_kind);
 
-    // Increase the number of frames in the current batch
-    curr_batch_counter_++;
+        // No sync needed here, the host doesn't need to wait for the copy to
+        // end. Only the consumer needs to be sure the data is here before
+        // manipulating it.
 
-    // The current batch is full
-    if (curr_batch_counter_ == batch_size_)
-    {
-        curr_batch_counter_ = 0;
-        const uint prev_end_index = end_index_;
-        end_index_ = (end_index_ + 1) % max_size_;
+        // Increase the number of frames in the current batch
+        curr_batch_counter_ += frames_to_enqueue;
 
-        // The queue is full (in terms of batch)
-        if (size_ == max_size_)
+        // The current batch is full
+        if (curr_batch_counter_ == batch_size_)
         {
-            has_overwritten_ = true;
-            start_index_ = (start_index_ + 1) % max_size_;
+            curr_batch_counter_ = 0;
+            const uint prev_end_index = end_index_;
+            end_index_ = (end_index_ + 1) % max_size_;
+
+            // The queue is full (in terms of batch)
+            if (size_ == max_size_)
+            {
+                has_overwritten_ = true;
+                start_index_ = (start_index_ + 1) % max_size_;
+            }
+            else
+            {
+                size_++;
+                curr_nb_frames_ += batch_size_;
+            }
+
+            // Unlock the current batch mutex
+            batch_mutexes_[prev_end_index].unlock();
+            // No batch are busy anymore
+            // End of critical section between enqueue (producer) & resize
+            // (consumer)
+            m_producer_busy_.unlock();
+
+            // block the worker after completing a batch to allow the resizing of the queue.
+            if (resize_in_progress_)
+            {
+                while (resize_in_progress_)
+                    continue;
+                return;
```

Comment:
Why return ? It will drop frames no ?

Reply:
Technicaly yes but the Queue gets clear when we resize anyway.